remote_attestation/rust/src/attestation.rs

```diff
@@ -15,238 +15,545 @@
+    pub fn attestation_init(
+        mut self,
+    ) -> anyhow::Result<(AttestationInit, AttestationEngine<Client, Attesting>)> {
+        let attestation_init = AttestationInit {
+            random: self.state.random.to_vec(),
+        };
+
+        // Update current transcript.
+        self.transcript
+            .append(&attestation_init)
+            .context("Couldn't append attestation init to transcript")?;
+
+        let key_negotiator = KeyNegotiator::create(KeyNegotiatorType::Client)
+            .context("Couldn't create key negotiator")?;
+        let attestation_engine = self
+            .transition(key_negotiator)
+            .context("Couldn't transition from Initializing to Attesting state")?;
+        Ok((attestation_init, attestation_engine))
     }
+}
 
-    /// Verifies remote attestation, negotiates session keys and creates client [`AeadEncryptor`].
+impl AttestationEngine<Server, Initializing> {
+    pub fn new(behavior: AttestationBehavior) -> Self {
+        Self {
+            participant: std::marker::PhantomData,
+            behavior,
+            state: Initializing::new(),
+            transcript: Transcript::new(),
+        }
+    }
+
+    /// Responds to `AttestationInit` message by creating a `ServerIdentity` message.
     ///
-    /// Client encryptor uses client session key for encryption and server session key for
-    /// decryption.
-    pub fn create_client_encryptor(
-        self,
-        peer_identity: &AttestationIdentity,
-    ) -> anyhow::Result<AeadEncryptor> {
-        // Verify peer attestation info.
-        let peer_attestation_info = &peer_identity.attestation_info;
-        self.attestation_type
-            .verify_attestation_info(&peer_attestation_info)
-            .context("Couldn't verify peer attestation info")?;
-
-        // Agree on session keys and create encryptor.
-        self.key_negotiator
-            .create_encryptor(peer_identity.public_key.as_ref(), EncryptorType::Client)
-            .context("Couldn't derive session key")
+    /// `ServerIdentity` message contains an ephemeral public key for negotiating session keys.
+    /// If self attestation is enabled this message also provides necessary information to perform
+    /// remote attestation.
+    ///
+    /// Transitions [`AttestationEngine`] state from [`Initializing`] to [`Attesting`] state.
+    pub fn process_attestation_init(
+        mut self,
+        attestation_init: &AttestationInit,
+    ) -> anyhow::Result<(ServerIdentity, AttestationEngine<Server, Attesting>)> {
+        // Create server identity message.
+        let key_negotiator = KeyNegotiator::create(KeyNegotiatorType::Server)
+            .context("Couldn't create key negotiator")?;
+        let ephemeral_public_key = key_negotiator
+            .public_key()
+            .context("Couldn't get ephemeral public key")?;
+        let server_identity = if self.behavior.contains_self_attestation() {
```

Comment:
Is it possible to have the creation of the `server_identity` as a method in `AttestationBehavior`, and only update `self.transcript` and `server_identity.transcript_signature` in this function?